librz/core/core.c

```diff
@@ -2647,68 +2647,59 @@ RZ_API void rz_core_fini(RzCore *c) {
 	rz_core_task_break_all(&c->tasks);
 	rz_core_task_join(&c->tasks, NULL, -1);
 	rz_core_wait(c);
-	/* TODO: it leaks badly */
-	// update_sdb (c);
 	//  avoid double free
-	rz_list_free(c->ropchain);
-	rz_event_free(c->ev);
-	free(c->cmdlog);
-	free(c->lastsearch);
+	RZ_FREE2(c->ropchain, rz_list_free);
+	RZ_FREE2(c->ev, rz_event_free);
+	RZ_FREE(c->cmdlog);
+	RZ_FREE(c->lastsearch);
 	RZ_FREE(c->cons->pager);
-	free(c->cmdqueue);
-	free(c->lastcmd);
-	free(c->stkcmd);
-	rz_list_free(c->visual.tabs);
-	free(c->block);
-	rz_core_autocomplete_free(c->autocomplete);
-
-	rz_list_free(c->gadgets);
-	rz_num_free(c->num);
-	// TODO: sync or not? sdb_sync (c->sdb);
-	// TODO: sync all dbs?
-	// rz_core_file_free (c->file);
-	// c->file = NULL;
+	RZ_FREE(c->cmdqueue);
+	RZ_FREE(c->lastcmd);
+	RZ_FREE(c->stkcmd);
+	RZ_FREE2(c->visual.tabs, rz_list_free);
+	RZ_FREE(c->block);
+	RZ_FREE2(c->autocomplete, rz_core_autocomplete_free);
+
+	RZ_FREE2(c->gadgets, rz_list_free);
+	RZ_FREE2(c->num, rz_num_free);
 	RZ_FREE(c->table_query);
-	rz_io_free(c->io);
-	rz_list_free(c->files);
-	rz_list_free(c->watchers);
-	rz_list_free(c->scriptstack);
+	RZ_FREE2(c->io, rz_io_free);
+	RZ_FREE2(c->files, rz_list_free);
+	RZ_FREE2(c->watchers, rz_list_free);
+	RZ_FREE2(c->scriptstack, rz_list_free);
 	rz_core_task_scheduler_fini(&c->tasks);
-	c->rcmd = rz_cmd_free(c->rcmd);
-	rz_list_free(c->cmd_descriptors);
-	c->analysis = rz_analysis_free(c->analysis);
-	rz_asm_free(c->rasm);
-	c->rasm = NULL;
-	c->print = rz_print_free(c->print);
-	c->bin = (rz_bin_free(c->bin), NULL);
-	c->lang = (rz_lang_free(c->lang), NULL);
-	c->dbg = (rz_debug_free(c->dbg), NULL);
-	rz_config_free(c->config);
+	RZ_FREE2(c->rcmd, rz_cmd_free);
```

Comment:
I like the idea of the macro, but I'm not sure my mind can easily map RZ_FREE2 == custom free+set-null. Maybe calling it `RZ_FREE_CUSTOM`?

Reply:
yes, sure.